Bad scaling of ion properties after initialization

See #326

test_ion_DSPMDE_prop_pack.py ......F                                                                                                                                 [100%]

================================================================================ FAILURES =================================================================================
___________________________________________________________________________ test_seawater_data ____________________________________________________________________________

    @pytest.mark.component
    def test_seawater_data():
        m = ConcreteModel()
        m.fs = FlowsheetBlock(default={'dynamic':False})
        m.fs.properties = DSPMDEParameterBlock(default={
            "solute_list": ["Ca_2+", "SO4_2-", "Na_+", "Cl_-", "Mg_2+"],
            "diffusivity_data": {("Liq", "Ca_2+"): 0.792e-9,
                                 ("Liq", "SO4_2-"): 1.06e-9,
                                 ("Liq", "Na_+"): 1.33e-9,
                                 ("Liq", "Cl_-"): 2.03e-9,
                                 ("Liq", "Mg_2+"): 0.706e-9},
            "mw_data": {"H2O": 18e-3,
                        "Na_+": 23e-3,
                        "Ca_2+": 40e-3,
                        "Mg_2+": 24e-3,
                        "Cl_-": 35e-3,
                        "SO4_2-": 96e-3},
            "stokes_radius_data": {"Na_+": 0.184e-9,
                                   "Ca_2+": 0.309e-9,
                                   "Mg_2+": 0.347e-9,
                                   "Cl_-": 0.121e-9,
                                   "SO4_2-": 0.230e-9},
            "charge": {"Na_+": 1,
                       "Ca_2+": 2,
                       "Mg_2+": 2,
                       "Cl_-": -1,
                       "SO4_2-": -2},
        })

        m.fs.stream = stream = m.fs.properties.build_state_block([0], default={'defined_state': True})

        mass_flow_in = 1 * pyunits.kg / pyunits.s
        feed_mass_frac = {'Na_+': 11122e-6,
                          'Ca_2+': 382e-6,
                          'Mg_2+': 1394e-6,
                          'SO4_2-': 2136e-6,
                          'Cl_-': 20300e-6}
        for ion, x in feed_mass_frac.items():
            mol_comp_flow = x * pyunits.kg / pyunits.kg * mass_flow_in / stream[0].mw_comp[ion]

            stream[0].flow_mol_phase_comp['Liq', ion].fix(mol_comp_flow)

        H2O_mass_frac = 1 - sum(x for x in feed_mass_frac.values())
        H2O_mol_comp_flow = H2O_mass_frac * pyunits.kg / pyunits.kg * mass_flow_in / stream[0].mw_comp['H2O']

        stream[0].flow_mol_phase_comp['Liq', 'H2O'].fix(H2O_mol_comp_flow)
        stream[0].temperature.fix(298.15)
        stream[0].pressure.fix(101325)

        stream[0].assert_electroneutrality(tol=1e-2)

        metadata = m.fs.properties.get_metadata().properties
        for v_name in metadata:
            getattr(stream[0], v_name)
        assert stream[0].is_property_constructed('conc_mol_phase_comp')

        assert_units_consistent(m)

        check_dof(m, fail_flag=True)

        calculate_scaling_factors(m)

        # check if any variables are badly scaled
        #       pre-initialize (all good)
        badly_scaled_var_list = list(badly_scaled_var_generator(m))
        [print(i[0],i[1]) for i in badly_scaled_var_list]
        assert len(badly_scaled_var_list) == 0

        stream.initialize()

        # check if any variables are badly scaled
        #   post-initialize (bad)
        badly_scaled_var_list = list(badly_scaled_var_generator(m))
        [print(i[0],i[1]) for i in badly_scaled_var_list]
>       assert len(badly_scaled_var_list) == 0
E       assert 6 == 0
E        +  where 6 = len([(<pyomo.core.base.var._GeneralVarData object at 0x000002EBFE35BBE0>, 0.0003820006676411202), (<pyomo.core.base.var._G...>, 0.0004064225570631957), (<pyomo.core.base.var._GeneralVarData object at 0x000002EBFDEEDD60>, 0.0009899800563660229)])

test_ion_DSPMDE_prop_pack.py:441: AssertionError
-------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------
fs.stream[0].mass_frac_phase_comp[Liq,Ca_2+] 0.0003820006676411202
fs.stream[0].conc_mol_phase_comp[Liq,Ca_2+] 0.000955149641847844
fs.stream[0].conc_mass_phase_comp[Liq,Ca_2+] 0.00038200131050573796
fs.stream[0].mole_frac_phase_comp[Liq,Ca_2+] 0.00017444201517142436
fs.stream[0].mole_frac_phase_comp[Liq,SO4_2-] 0.0004064225570631957
fs.stream[0].molality_comp[Ca_2+] 0.0009899800563660229

---------- coverage: platform win32, python 3.8.12-final-0 -----------
Name                                                                                                                                                              Stmts   Miss  Cover
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
100 durations ==========================================================================
0.96s call     watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_property_ions
0.27s call     watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_seawater_data
0.09s setup    watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_scaling
0.04s call     watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_scaling
0.02s setup    watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_parameter_block
0.02s setup    watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_default_scaling
0.02s setup    watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_build
0.02s setup    watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_property_ions
0.02s setup    watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_general_methods
0.02s call     watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py::test_build

(11 durations < 0.005s hidden.  Use -vv to show these durations.)
========================================================================= short test summary info =========================================================================
FAILED test_ion_DSPMDE_prop_pack.py::test_seawater_data - assert 6 == 0
======================================================================= 1 failed, 6 passed in 6.45s =======================================================================

--- a/watertap/property_models/ion_DSPMDE_prop_pack.py
+++ b/watertap/property_models/ion_DSPMDE_prop_pack.py
@@ -672,12 +672,12 @@ class DSPMDEStateBlockData(StateBlockData):
 
         # these variables should have user input
         if iscale.get_scaling_factor(self.flow_mol_phase_comp['Liq', 'H2O']) is None:
-            sf = iscale.get_scaling_factor(self.flow_mol_phase_comp['Liq', 'H2O'], default=1e0)
+            sf = iscale.get_scaling_factor(self.flow_mol_phase_comp['Liq', 'H2O'], default=1, warning=True)
             iscale.set_scaling_factor(self.flow_mol_phase_comp['Liq', 'H2O'], sf)
 
         for j in self.params.solute_set:
             if iscale.get_scaling_factor(self.flow_mol_phase_comp['Liq', j]) is None:
-                sf = iscale.get_scaling_factor(self.flow_mol_phase_comp['Liq', j], default=1)
+                sf = iscale.get_scaling_factor(self.flow_mol_phase_comp['Liq', j], default=1, warning=True)
                 iscale.set_scaling_factor(self.flow_mol_phase_comp['Liq', j], sf)
 
         # scaling factors for parameters

--- a/watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py
+++ b/watertap/property_models/tests/test_ion_DSPMDE_prop_pack.py
@@ -424,9 +424,23 @@ def test_seawater_data():
 
     check_dof(m, fail_flag=True)
 
+    m.fs.properties.set_default_scaling('flow_mol_phase_comp', 1, index=('Liq', 'H2O'))
+    m.fs.properties.set_default_scaling('flow_mol_phase_comp', 1, index=('Liq', 'Na_+'))
+    m.fs.properties.set_default_scaling('flow_mol_phase_comp', 1, index=('Liq', 'Cl_-'))
+    m.fs.properties.set_default_scaling('flow_mol_phase_comp', 1e1, index=('Liq', 'Ca_2+'))
+    m.fs.properties.set_default_scaling('flow_mol_phase_comp', 1e1, index=('Liq', 'SO4_2-'))
+    m.fs.properties.set_default_scaling('flow_mol_phase_comp', 1, index=('Liq', 'Mg_2+'))
     calculate_scaling_factors(m)
 
+    # check if any variables are badly scaled
+    badly_scaled_var_list = list(badly_scaled_var_generator(m))
+    assert len(badly_scaled_var_list) == 0
+
     stream.initialize()
+
+    # check if any variables are badly scaled
+    badly_scaled_var_list = list(badly_scaled_var_generator(m))
+    assert len(badly_scaled_var_list) == 0
 
     results = solver.solve(m)
     assert_optimal_termination(results)

